### DISCUSSION TITLE
CSRF Form Issue

### DISCUSSION THREAD

**Discussion by meccoworld** [phalcon/cphalcon#16321] [Q&A]
I have been trying to set up a CSRF form with the below code but checkToken() always return false.
```
        if ($this->request->isPost() && $this->security->checkToken()) {
            $this->view->message = "Okay";
        }else{
            $this->view->message = "Not Okay";
        }
```

**Comment by noone-silent** [comment]
Do you have any Ajax requests on this page? Also check if all resources (images etc.) are loaded correctly (no HTTP Status 404). It could be that this could trigger a new request to Phalcon and generate a new Token.

**Reply by meccoworld** [reply]
Hi noone,

Thanks for the response, below is my code.

```php
<head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta name="robots" content="noindex">
<link rel="shortcut icon" type="image/x-icon" href="">
<?php $this->assets->outputCss(); ?>
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css">
</head>


<form action="signup/register" method="post">
<input type="hidden" name="<?php echo $this->security->getTokenKey() ?>" value="<?php echo $this->security->getToken() ?>"/>
``` 




```php



 public function registerAction()
    {

        if ($this->request->isPost()) {
            if ($this->security->checkToken()) {
                // The token is OK
                return "CSRF Working.";
            }else{
                return "CSRF not Working.";
            }
        }

``` 

When I check view source from the registration page, the token input name and the values gets generated, but when I submit it doesn't validate.

Thank you,
Kind regards

**Reply by noone-silent** [reply]
I mean like this:

1. open your brower
2. Press F12 or open the developer tools over the menu
3. Open the Network Tab
4. Open the Site where the bug happens
5. Look in the Network Tab if some requests fail (Status Code 404, 500 etc.)

For example the following can happen. You call an image or css or font whatever, but the resource is not found anymore and the request will be routed through the Phalcon framework. With this, a new request has been made and a new token will be generated. If you now submit the form, the token do not match anymore.

**Comment by meccoworld** [comment]
I have check everything seems to return 200, please see attachment of screenshots

<img width="1440" alt="Screenshot 2023-04-20 at 8 14 06 AM" src="https://user-images.githubusercontent.com/15078022/233289200-c94e086e-fc18-46af-848c-e86e49e2d479.png">


<img width="1440" alt="Screenshot 2023-04-20 at 8 14 16 AM" src="https://user-images.githubusercontent.com/15078022/233289252-019b3bd3-bfc1-46ba-a69c-6a97a55f31e4.png">

**Comment by noone-silent** [comment]
Could you please check if the 2 css files have the right content? You can click them in the network tab and see the contents

**Reply by noone-silent** [reply]
And you have 7 errors in the console tab

**Comment by meccoworld** [comment]
Hello noone,
Still not working, I even comment out the CSS files and CSRF still now working.
Thank you

**Comment by noone-silent** [comment]
Do you have a running session? CSRF needs a running session where it can store the token, also try to log every call in the `index.php` then you will see if the script gets called from somewhere else.

**Reply by meccoworld** [reply]
I’m not sure how I can do that.

But I’m sure the csrf input name and values were generated from session so I would say I have a session running but not sure about where it can store token.

Thanks

**Reply by noone-silent** [reply]
You can check if the session is started. See https://docs.phalcon.io/4.0/en/session#dependency-injection

There can only be 2 problems here: Session doesn't start or save the token or another request in the background forces the Framework to generate a new Token (because every Token is only valid for 1 request).

To log every request that is happening you could do something like this in the `index.php`:

```php
file_put_contents('request.log', date('Y-m-d H:i:s') . ' ' . $_SERVER['REQUEST_URI'] . \PHP_EOL, \FILE_APPEND);
```

create a `request.log` file next to the `index.php` and give write permissions.

Do one request and look into the `request.log`. If there are 2 line in this file, you know there have been another request. If there is only 1 line, then the problem would be the session.

**Comment by meccoworld** [comment]
Please kindly see the result from the above execution.

My session is started. If it is okay I don't know if we can connect for a Zoom meeting because this is really taking long for me to proceed to my next steps.

<img width="860" alt="Screenshot 2023-04-21 at 8 37 36 AM" src="https://user-images.githubusercontent.com/15078022/233574408-1bb28a79-0378-413a-9b48-7f559aad159c.png">

<img width="1323" alt="Screenshot 2023-04-21 at 8 39 44 AM" src="https://user-images.githubusercontent.com/15078022/233574512-1c8a8229-df48-42fb-97c5-bfa291a45c05.png">


Thank you

**Reply by noone-silent** [reply]
the request log needs to be in `public/index.php` to track every request. Here it only tracks a call to `indexAction`.

Also please make sure you have an active session handler registered in your services. Please see https://docs.phalcon.io/5.0/en/session#dependency-injection and add that to your service definitions.

**Comment by meccoworld** [comment]
I noticed a change.

<img width="985" alt="Screenshot 2023-04-21 at 9 35 18 AM" src="https://user-images.githubusercontent.com/15078022/233588405-c16f08e5-4f96-4c4b-999a-c13437f0663a.png">

<img width="1150" alt="Screenshot 2023-04-21 at 9 35 24 AM" src="https://user-images.githubusercontent.com/15078022/233588433-48d33fba-d387-4788-8ee4-a4523a43a6b2.png">

**Comment by meccoworld** [comment]
It working now. Thanks.

I had to use the below HTML code to prevent making additional favicon requests.

`<link rel="shortcut icon" href="data:image/x-icon;," type="image/x-icon"> `

**Comment by noone-silent** [comment]
That's what I meant. The `favicon.ico` does not exist. Then it will be forwarded to the Framework and a new Token will be generated.

**Reply by meccoworld** [reply]
Thank you. I appreciate the help.

I hope you don't mind when I come back with other challenges later.

Thanks again.
